donkeycar/parts/lidar.py

```diff
@@ -13,6 +14,187 @@
 from donkeycar.utils import norm_deg, dist, deg2rad, arr_to_img
 from PIL import Image, ImageDraw
 
+
+CLOCKWISE = 1
+COUNTER_CLOCKWISE = -1
+
+
+class RPLidar2(object):
+    '''
+    Adapted from https://github.com/Ezward/rplidar
+    '''
+    def __init__(self,
+                 min_angle = 0.0, max_angle = 360.0,
+                 min_distance = sys.float_info.min, max_distance = sys.float_info.max,
+                 forward_angle = 0.0,
+                 angle_direction=CLOCKWISE,
+                 debug=False):
+        
+        self.lidar = None
+        self.port = None
+        self.on = False
+        
+
+        help = []
+        if min_distance < 0:
+            help.append("min_distance must be >= 0")
+
+        if max_distance <= 0:
+            help.append("max_distance must be > 0")
+            
+        if min_angle < 0 or min_angle > 360:
+            help.append("min_angle must be 0 <= min_angle <= 360")
+
+        if max_angle <= 0 or max_angle > 360:
+            help.append("max_angle must be 0 < max_angle <= 360")
+          
+        if forward_angle < 0 or forward_angle > 360:
+            help.append("forward_angle must be 0 <= forward_angle <= 360")
+            
+        if angle_direction != CLOCKWISE and angle_direction != COUNTER_CLOCKWISE:
+            help.append("angle-direction must be 1 (clockwise) or -1 (counter-clockwise)")
+
+        if len(help) > 0:
+            msg = "Could not start RPLidar; bad parameters passed to constructor"
+            print(msg)
+            for h in help:
+                print("  " + h)
+            raise ValueError(msg)
+
+        self.min_angle = min_angle
+        self.max_angle = max_angle
+        self.min_distance = min_distance
+        self.max_distance = max_distance
+        self.forward_angle = forward_angle
+        self.spin_reverse = (args.angle_direction != CLOCKWISE)
+        self.measurements = [] #a list of (angle, distance, time, scan, index) measurements
+
+        from adafruit_rplidar import RPLidar
+        import glob
+        
+        #
+        # find the serial port where the lidar is connected
+        #
+        port_found = False
+        temp_list = glob.glob ('/dev/ttyUSB*')
+        result = []
+        for a_port in temp_list:
+            try:
+                s = serial.Serial(a_port)
+                s.close()
+                result.append(a_port)
+                port_found = True
+            except serial.SerialException:
+                pass
```

Comment:
As you are not using the context manager, don't you need the `close()` of the serial stream here? Probably change to context manager.

Reply:
I thought the same thing when I looked at the original code (I just copied this from the older part).  The thing is, we only open the serial port, then immediately close it.  We just want to know if it is available.  If it won't open, we get an exception.  If it does open we immediately close it.  So the context manager doesn't make sense here; we are not doing any work with the serial port.  In terms of if the code should call serial close; if we get an exception, we never got a serial port to close, so we can't close it.

Reply:
👍